src/main/java/com/google/edith/services/StoreReceiptService.java

```diff
@@ -0,0 +1,159 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.edith.services;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.edith.interfaces.StoreReceiptInterface;
+import com.google.edith.servlets.Item;
+import com.google.edith.servlets.Receipt;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Optional;
+import javax.servlet.http.HttpServletRequest;
+
+public final class StoreReceiptService implements StoreReceiptInterface {
+
+  private final DatastoreService datastore;
+
+  public StoreReceiptService(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  @Override
+  public void storeEntites(Receipt receipt) {
+    storeReceiptEntity(receipt);
+  }
+
+  @Override
+  public Receipt parseReceiptFromForm(HttpServletRequest request) throws IOException {
+    BufferedReader bufferedReader = request.getReader();
+    Gson gson = new GsonBuilder().registerTypeAdapter(Item.class, new ItemDeserializer()).create();
+    JsonParser parser = new JsonParser();
+    JsonObject json = (JsonObject) parser.parse(bufferedReader);
+    return gson.fromJson(json, Receipt.class);
+  }
```

Comment:
This method should be static.

This method should be in the servlet since it deals with HttpServletRequest; however, you can also make this method static and remove it from the interface, which would make the interface much more portable by decoupling it from the HttpServlet technology.

Reply:
decoupled using ```java.io.BufferedReader;```